docs/internal_abi.md

```diff
@@ -0,0 +1,321 @@
+
+It is therefore possible to fully encode this example in a single 64-bit word on any common 64-bit system.
+For example, in userspace on a little-endian 64-bit system with a 48-bit virtual address space that uses integers as pointers:
+
+ - If any of bits 50-62 inclusive is non-zero, then the entire value is a `Float64`.
+ - Otherwise, if the value of bit 0 is 1, then bits 8-15 inclusive contain a `U8` value.
+ - Otherwise, bits 0-47 contain should be zero extended to give a 64-bit address.
+
+This form of encoding is one of the motivations for supporting storage-only types such as `U42`.
+Any type that does not use the full width of a machine word is amenable to dense packing in discriminated unions.
+This is a sufficiently common pattern in systems code that Verona has chosen to bake it into the language.
+
+The rules for determining the correct encoding for a union type begins by identifying the number of payload bits and the number of states required for the type discriminator:
+
+ - For singleton types, there are two possible answers:
+   - Each type requires one state in the discriminator but no bits to encode the value.
+   - Each type requires one pointer's worth of state to encode the value but no bits in the discriminator.
+   We choose the former, to maximise our ability to use the CPU's branch-on-zero instructions.
+ - Each integer requires one state in the discriminator and as many bits as its width defines for the payload.
+   For example, in the union `U8 | U16`, each requires one state in the discriminator (giving a 1-bit discriminator to represent both state) and 8 and 16 bits of payload, respectively.
+ - Floating point types follow the same rules as integers, with special handling if a floating point type is the type with the largest payload in a union type, described below.
+ - Pointers to object types (including external reference types) require as many bits as are required for an unambiguous pointer; however, all pointer types require only a single discriminator state because all pointees can be discriminated by inspecting their descriptor.
+   The number of bits required for a pointer varies depending on the pointer representation:
+   - 32-bit integer pointers require 29 bits of state.
+   - 64-bit integer pointers require either 45 or 52 bits, depending on the virtual address space size of the target.
+   - CHERI capabilities require 125 bits of state but provide a 'free' discriminator bit because the tag bit can be used to discriminate CHERI capabilities from non-pointer data.
+ - Value types are still being defined; however, it seems plausible that they will each require one descriptor state and as many payload bits as it has non-padding bits.
+   Making efficient use of padding bits as discriminators seems very difficult and so is likely to be left to a later ABI, if ever implemented.
+
+Once the payload sizes are all known, the algorithm for implementing the packing depends on the type of the largest payload.
```

Comment:
I worry about the performance differences of saying `U16 | F32` versus `U32 | F16`. It's not entirely clear to the programmer model that they will be encoded completely differently, which could have impact on memory and run time.

Reply:
Hopefully if people need to care, they are able to read the docs on how the encoding works.  Lazy people can assume that the encoding of both is the same as:
```c
struct
{
  union
  {
      float f;
      short i;
  };
  int discriminator:1;
};
```
Or
```c
struct
{
  union
  {
      fp16 f;
      int i;
  };
  int discriminator:1;
};
```
Anything else is an optimisation.

Reply:
Actually, it wouldn't be bad to have at least one of those examples in the document.

What I was worried about is that we do such a good job at the optimisation side for some and not others, the "others" will look like we did a bad job. Like Java's Integer caches.